parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+        return Output(preds, cand_preds)
+
+    def observe(self, observation):
+        labels = observation.get('labels')
+        if not labels:
+            labels = observation.get('eval_labels')
+        if labels:
+            for label in labels:
+                self.add_to_ys_cache(label)
+        return super(StarspaceTorchAgent, self).observe(observation)
+
+    def vectorize(self, obs, add_start=False, add_end=False, truncate=None,
+                  split_lines=False):
+        """Overrride method from Torch Agent so that by default we do not add
+        start and end indices."""
+        return super(StarspaceTorchAgent, self).vectorize(
+            obs,
+            add_start=add_start,
+            add_end=add_end,
+            truncate=truncate,
+            split_lines=split_lines,
+        )
+
+    def _get_custom_candidates(self, obs):
+        """Sets custom candidates for the observation."""
+        if not obs.get('eval_labels'):
+            # only use custom candidates during training
+            cands = []
+            cache_sz = len(self.ys_cache) - 1
+            if cache_sz < 1:
+                return cands
+            k = self.opt['neg_samples']
+            for i in range(1, k * 3):
+                index = random.randint(0, cache_sz)
+                cand = self.ys_cache[index]
+                if not self.same(obs.get('labels'), cand):
+                    cands.append(cand)
+                    if len(cands) >= k:
+                        break
+            if self.opt['parrot_neg']:
+                query = self.history.split('\n')[-1]
+                cands.append(query)
+                return cands
+        # if we aren't training, return label candidates if available
+        return obs.get('label_candidates')
+
+    def _init_embeddings(self, log=True):
+        """Copy embeddings from the pretrained embeddings to the lookuptable.
+
+        :param weight:   weights of lookup table (nn.Embedding/nn.EmbeddingBag)
+        :param emb_type: pretrained embedding type
+        """
+        weight = self.model.lt.weight
+        emb_type = self.opt.get('embedding_type', 'random')
+        if emb_type == 'random':
+            return
+        embs, name = self._get_embtype(self, emb_type)
+        cnt = 0
+        for w, i in self.dict.tok2ind.items():
+            if w in embs.stoi:
+                vec = self._project_vec(self,
+                                        embs.vectors[embs.stoi[w]],
+                                        weight.size(1))
+                weight.data[i] = vec
+                cnt += 1
+        if log:
+            print('Initialized embeddings for {} tokens ({}%) from {}.'
+                  ''.format(cnt, round(cnt * 100 / len(self.dict), 1), name))
+
+    def same(self, labels, cand):
+        if labels:
+            for label in labels:
+                if label == cand:
+                    return True
+        return False
+
+    def input_dropout(self, xs, ys, negs):
```

Comment:
Hmm, this looks like a variant on UnknownDropout in Seq2Seq. I wonder if it's worth refactoring both of these. https://github.com/facebookresearch/ParlAI/blob/9e113105d1690deda0b372f6aeaf933ab58e3a72/parlai/agents/seq2seq/modules.py#L149-L172

Reply:
yep, it's a bit different. namely, in starspace we are getting rid of random tokens in the sentence at some rate and here we are replacing them with unknown tokens (the latter is a bit cleaner).. i'll think about ways to make this better